### PR TITLE
Implement method kj::StringPtr:parseAs<NUMBER_TYPE>

### PR DESCRIPTION
**Pull request by katreniak**
This  PR moves number parsing code from capnp json decoder to kj::StringPtr as discussed [here](https://github.com/sandstorm-io/capnproto/pull/297#issuecomment-212764806).

Questions:

1. Integer parsing methods use base 10. My problem with auto detected base is that 0 prefixed numbers are correct decimal numbers but they are interpreted as octal. Does anybody use octal numbers? I would be ok to automatically interpret hexadecimal numbers prefixed with "0x" or "0X". 
2. There is no optional base parameter to integer parsing. I am not sure how to write the template magic that accepts optional argument only for integral types. Is this important / necessary?
3. There is one more call to `strtod` in json parser. But it throws when HUGE floating number overflows instead of interpreting the number as INF. Current `parseAs<double>()` code parses it as INF. What is the desired behavior for each of them?

@kentonv @kamalmarhubi

### CODE REVIEW

File: c++/src/kj/string.c++ [hunk 1]
```diff
@@ -33,6 +34,71 @@ namespace kj {
...
+  KJ_REQUIRE(errno != ERANGE, "Value out-of-range", s);
+  KJ_REQUIRE(value <= max, "Value out-of-range", value, max);
+  KJ_REQUIRE(s[0] != '-', "Value out-of-range", s); //strtoull("-1") does not fail with ERANGE
+  return value;
+}
+
+template <typename T>
+T parseInteger(const StringPtr& s) {
+  if (std::numeric_limits<T>::is_signed) {
+    auto min = (long long) std::numeric_limits<T>::min();
```
Comment (kentonv) [inline c++/src/kj/string.c++]: You could avoid pulling in `std::numeric_limits` here by using `minValue` and `maxValue`.

    if (static_cast<T>(minValue) < 0) {
      long long min = static_cast<T>(minValue);
      long long max = static_cast<T>(maxValue);

**Comment by katreniak** [comment]
* removed #include <limits>
* added support for hex numbers
* added documentation

File: c++/src/kj/string.c++ [hunk 2]
```diff
@@ -61,12 +60,12 @@ unsigned long long parseUnsigned(const StringPtr& s, unsigned long long max) {
...
-  if (std::numeric_limits<T>::is_signed) {
-    auto min = (long long) std::numeric_limits<T>::min();
-    auto max = (long long) std::numeric_limits<T>::max();
+  if (static_cast<T>(minValue) < 0) {
+    long long min = static_cast<T>(minValue);
+    long long max = static_cast<T>(maxValue);
     return static_cast<T>(parseSigned(s, min, max));
   } else {
-    auto max = (unsigned long long) std::numeric_limits<T>::max();
+    long long max = static_cast<T>(maxValue);
```
Comment (kentonv) [inline c++/src/kj/string.c++]: This should be `unsigned long long`, right?

Comment (katreniak) [inline c++/src/kj/string.c++]: True. Fixed in 651aaf3b5def9808c3272fb5565122a3921f9aa2

**Comment by kentonv** [comment]
Thanks!

**Pull request merged**